File: app/src/main/java/com/savvasdalkitsis/gameframe/grid/model/ColorGrid.java
package com.savvasdalkitsis.gameframe.grid.model;

import android.graphics.Color;
import android.support.annotation.ColorInt;

import com.savvasdalkitsis.gameframe.composition.model.ARGB;
import com.savvasdalkitsis.gameframe.composition.model.BlendMode;
import com.savvasdalkitsis.gameframe.composition.model.PorterDuffOperator;
import com.savvasdalkitsis.gameframe.composition.usecase.BlendUseCase;
import com.savvasdalkitsis.gameframe.injector.usecase.UseCaseInjector;

public class ColorGrid {

    public static final int SIDE = 16;
    private final int[][] colors = new int[SIDE][SIDE];
    private final BlendUseCase blendUseCase = UseCaseInjector.blendUseCase();

    public ColorGrid() {
        fill(Color.TRANSPARENT);
    }

    public void setColor(@ColorInt int color, int column, int row) {
        checkValue(column, "Column");
        checkValue(row, "Row");
        colors[column - 1][row -1] = color;
    }

    public void fill(int color) {
        for (int i = 1; i <= SIDE; i++) {
            for (int j = 1; j <= SIDE; j++) {
                setColor(color, i, j);
            }
        }
    }

    @ColorInt
    public int getColor(int column, int row) {
        checkValue(column, "Column");
        checkValue(row, "Row");
        return colors[column - 1][row - 1];
    }

    private void checkValue(int value, final String valueName) {
        if (value < 1 || value > SIDE) {
            throw new IllegalArgumentException(valueName + " value should be between 1 and 16");
        }
    }

    public boolean isOutOfBounds(int column, int row) {
        return row < 1 || column < 1 || row > SIDE || column > SIDE;
    }

    public ColorGrid compose(ColorGrid dest, ColorGrid source, BlendMode blendMode, PorterDuffOperator porterDuffOperator, float alpha) {
        ColorGrid colorGrid = new ColorGrid();
        for (int col = 1; col <= SIDE; col++) {
            for (int row = 1; row <= SIDE; row++) {
                ARGB blend = blendUseCase.blend(source.getColor(col, row), dest.getColor(col, row),
                        blendMode, porterDuffOperator, alpha);
                colorGrid.setColor(blend.color(), col, row);
            }
        }
        return colorGrid;
    }
}

File: app/src/main/java/com/savvasdalkitsis/gameframe/feature/workspace/element/grid/model/ColorGrid.kt
package com.savvasdalkitsis.gameframe.feature.workspace.element.grid.model

import android.graphics.Color
import android.support.annotation.ColorInt
import com.savvasdalkitsis.gameframe.feature.composition.model.AvailableBlendMode
import com.savvasdalkitsis.gameframe.feature.composition.model.AvailablePorterDuffOperator
import com.savvasdalkitsis.gameframe.infra.kotlin.clip
import com.savvasdalkitsis.gameframe.injector.usecase.UseCaseInjector

class ColorGrid : Grid {

    private val colors = Array(SIDE) { IntArray(SIDE) }
    private var translateCol: Int = 0
    private var translateRow: Int = 0

    @Transient
    private val blendUseCase = UseCaseInjector.blendUseCase()
    @Transient
    private var transientTranslateCol: Int = 0
    @Transient
    private var transientTranslateRow: Int = 0
    @Transient
    private var scratch: ColorGrid? = null

    init {
        fill(Color.TRANSPARENT)
    }

    override fun setColor(@ColorInt color: Int, column: Int, row: Int) {
        checkValue(column, "Column")
        checkValue(row, "Row")
        val c = column - columnTranslation
        val r = row - rowTranslation
        if (!isOutOfBounds(c, r)) {
            colors[c - 1][r - 1] = color
        }
    }

    override fun fill(color: Int): ColorGrid {
        for (i in 1..SIDE) {
            for (j in 1..SIDE) {
                setColor(color, i, j)
            }
        }
        return this
    }

    @ColorInt
    override fun getColor(column: Int, row: Int): Int {
        checkValue(column, "Column")
        checkValue(row, "Row")
        val c = column - columnTranslation
        val r = row - rowTranslation
        if (isOutOfBounds(c, r)) {
            return Color.TRANSPARENT
        }
        val color = colors[c - 1][r - 1]
        return if (scratch == null) {
            color
        } else {
            blendUseCase.mix(scratch!!.getColor(column, row), color, AvailableBlendMode.NORMAL, AvailablePorterDuffOperator.SOURCE_OVER, 1f).color()
        }
    }

    override fun replicateMoment(): ColorGrid {
        val colorGrid = ColorGrid()
        colorGrid.copyColorsFrom(this)
        colorGrid.translateCol = translateCol
        colorGrid.translateRow = translateRow
        return colorGrid
    }

    override fun isIdenticalTo(moment: Grid): Boolean {
        if (moment !is ColorGrid) {
            return false
        }
        if (translateCol != moment.translateCol || translateRow != moment.translateRow) {
            return false
        }
        for (i in 0 until SIDE) {
            for (j in 0 until SIDE) {
                if (colors[i][j] != moment.colors[i][j]) {
                    return false
                }
            }
        }
        return true
    }

    private fun checkValue(value: Int, valueName: String) {
        if (value < 1 || value > SIDE) {
            throw IllegalArgumentException("$valueName value should be between 1 and $SIDE but was $value")
        }
    }

    fun translate(translateCol: Int, translateRow: Int) {
        transientTranslateCol = translateCol
        transientTranslateRow = translateRow
    }

    fun freezeTranslation() {
        translateCol = bound(columnTranslation)
        translateRow = bound(rowTranslation)
        transientTranslateCol = 0
        transientTranslateRow = 0
    }

    val rowTranslation: Int
        get() = translateRow + transientTranslateRow

    val columnTranslation: Int
        get() = translateCol + transientTranslateCol

    fun initializeScratch(): Grid {
        scratch = ColorGrid()
        return scratch as ColorGrid
    }

    fun rasterScratch() {
        if (scratch != null) {
            val scratch = this.scratch as ColorGrid
            this.scratch = null
            val grid = blendUseCase.compose(this, scratch, AvailableBlendMode.NORMAL, AvailablePorterDuffOperator.SOURCE_OVER, 1f)
            copyColorsFrom(grid)
        }
    }

    private fun copyColorsFrom(grid: ColorGrid) {
        for (i in 0 until SIDE) {
            System.arraycopy(grid.colors[i], 0, colors[i], 0, SIDE)
        }
    }

    private fun bound(value: Int): Int = value.clip(-SIDE, SIDE)

    companion object {

        const val SIDE = 16

        fun isOutOfBounds(column: Int, row: Int): Boolean {
            return row < 1 || column < 1 || row > SIDE || column > SIDE
        }
    }
}
